crates/bevy_input_focus/src/directional_navigation.rs

```diff
@@ -259,8 +363,323 @@ pub enum DirectionalNavigationError {
+/// ];
+///
+/// auto_generate_navigation_edges(&mut nav_map, &nodes, &config);
+/// ```
+pub fn auto_generate_navigation_edges(
+    nav_map: &mut DirectionalNavigationMap,
+    nodes: &[NavigationNode],
+    config: &AutoNavigationConfig,
+) {
+    // For each node, find best neighbor in each direction
+    for origin in nodes {
+        for octant in [
+            CompassOctant::North,
+            CompassOctant::NorthEast,
+            CompassOctant::East,
+            CompassOctant::SouthEast,
+            CompassOctant::South,
+            CompassOctant::SouthWest,
+            CompassOctant::West,
+            CompassOctant::NorthWest,
+        ] {
+            // Skip if manual edge already exists (check inline to avoid borrow issues)
+            if nav_map
+                .get_neighbors(origin.entity)
+                .and_then(|neighbors| neighbors.get(octant))
+                .is_some()
+            {
+                continue; // Respect manual override
+            }
+
+            // Find best candidate in this direction
+            let mut best_candidate = None;
+            let mut best_score = f32::INFINITY;
+
+            for candidate in nodes {
+                // Skip self
+                if candidate.entity == origin.entity {
+                    continue;
+                }
+
+                // Score the candidate
+                let score = score_candidate(
+                    origin.position,
+                    origin.size,
+                    candidate.position,
+                    candidate.size,
+                    octant,
+                    config,
+                );
+
+                if score < best_score {
+                    best_score = score;
+                    best_candidate = Some(candidate.entity);
+                }
+            }
+
+            // Add edge if we found a valid candidate
+            if let Some(neighbor) = best_candidate {
+                nav_map.add_edge(origin.entity, neighbor, octant);
+            }
+        }
+    }
+}
+
+/// Built-in system that automatically rebuilds the navigation graph for `bevy_ui` nodes.
+///
+/// This system runs in `PostUpdate` after `UiSystems::Layout` and automatically updates
+/// the navigation graph when nodes with [`AutoDirectionalNavigation`] component change
+/// their position or size.
+///
+/// # How it works
+///
+/// 1. Detects nodes with [`AutoDirectionalNavigation`] that have changed
+/// 2. Extracts position/size from [`ComputedNode`] and [`UiGlobalTransform`]
+/// 3. Calls [`auto_generate_navigation_edges`] to rebuild connections
+///
+/// This system is automatically added by [`DirectionalNavigationPlugin`], so users
+/// only need to add the [`AutoDirectionalNavigation`] component to their UI entities.
+///
+/// # Note
+///
+/// This system only works with `bevy_ui` nodes. For custom UI systems, call
+/// [`auto_generate_navigation_edges`] directly in your own system.
+fn auto_rebuild_ui_navigation_graph(
```

Comment:
This system needs to account for visibility, and not suggest that users navigate to nodes that are currently set as invisible.